flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -0,0 +1,233 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.resourcemanager.slotmanager;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.clusterframework.types.ResourceProfile;
+import org.apache.flink.runtime.instance.InstanceID;
+import org.apache.flink.runtime.slots.ResourceCounter;
+import org.apache.flink.runtime.slots.ResourceRequirement;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.runtime.resourcemanager.slotmanager.SlotManagerUtils.getEffectiveResourceProfile;
+
+/**
+ * The default implementation of {@link ResourceAllocationStrategy}, which always allocate pending
+ * task managers with the fixed profile.
+ */
+public class DefaultResourceAllocationStrategy implements ResourceAllocationStrategy {
+    private final ResourceProfile defaultSlotResourceProfile;
+    private final ResourceProfile totalResourceProfile;
+
+    public DefaultResourceAllocationStrategy(
+            ResourceProfile defaultSlotResourceProfile, int numSlotsPerWorker) {
+        this.defaultSlotResourceProfile = defaultSlotResourceProfile;
+        this.totalResourceProfile = defaultSlotResourceProfile.multiply(numSlotsPerWorker);
+    }
+
+    /**
+     * Matches resource requirements against available and pending resources. For each job, in a
+     * first round requirements are matched against registered resources. The remaining unfulfilled
+     * requirements are matched against pending resources, allocating more workers if no matching
+     * pending resources could be found. If the requirements for a job could not be fulfilled then
+     * it will be recorded in {@link ResourceAllocationResult#getUnfulfillableJobs()}.
+     *
+     * <p>Performance notes: At it's core this method loops, for each job, over all resources for
+     * each required slot, trying to find a matching registered/pending task manager. One should
+     * generally go in with the assumption that this runs in numberOfJobsRequiringResources *
+     * numberOfRequiredSlots * numberOfFreeOrPendingTaskManagers.
+     *
+     * <p>In the absolute worst case, with J jobs, requiring R slots each with a unique resource
+     * profile such each pair of these profiles is not matching, and T registered/pending task
+     * managers that don't fulfill any requirement, then this method does a total of J*R*T resource
+     * profile comparisons.
+     */
```

Comment:
The computation is way too heavy at the moment.
Given that this PR is already quite big, I'm ok with it as the first step. But we should open a follow-up ticket to optimize it.

Reply:
I open https://issues.apache.org/jira/browse/FLINK-21174 for it.